pkg/ottl/context_inferrer.go

```diff
@@ -0,0 +1,85 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottl // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+
+import "math"
+
+var (
+	defaultContextInferPriority = []string{
+		"log",
+		"metric",
+		"datapoint",
+		"spanevent",
+		"span",
+		"resource",
+		"scope",
+		"instrumentation_scope",
+	}
+)
+
+// ContextInferrer is an interface used to infer the OTTL context from statements paths.
+type ContextInferrer interface {
+	// Infer returns the OTTL context inferred from the given statements paths.
+	Infer(statements []string) (string, error)
+}
+
+type priorityContextInferrer struct {
+	contextPriority      map[string]int
+	ignoreUnknownContext bool
+}
+
+func (s *priorityContextInferrer) Infer(statements []string) (string, error) {
+	var inferredContext string
+	var inferredContextPriority int
+
+	for _, statement := range statements {
+		parsed, err := parseStatement(statement)
+		if err != nil {
+			return inferredContext, err
+		}
+
+		for _, p := range getParsedStatementPaths(parsed) {
+			pathContextPriority, ok := s.contextPriority[p.Context]
+			if !ok {
+				if s.ignoreUnknownContext {
+					continue
+				}
+
+				// Lowest priority
+				pathContextPriority = math.MaxInt
+			}
+
+			if inferredContext == "" || pathContextPriority < inferredContextPriority {
+				inferredContext = p.Context
+				inferredContextPriority = pathContextPriority
+			}
+		}
+	}
+
+	return inferredContext, nil
+}
+
+// DefaultPriorityContextInferrer is like NewPriorityContextInferrer, but using the default
+// context priorities and ignoring unknown/non-prioritized contexts.
+func DefaultPriorityContextInferrer() ContextInferrer {
+	return NewPriorityContextInferrer(defaultContextInferPriority, false)
+}
+
+// NewPriorityContextInferrer creates a new priority-based context inferrer.
+// To infer the context, it compares all [ottl.Path.Context] values, prioritizing them based
+// on the provide contextsPriority argument, the lower the context position is in the array,
+// the more priority it will have over other items.
+// If unknown/non-prioritized contexts are found on the statements, they can be either ignored
+// or considered when no other prioritized context is found. To skip unknown contexts, the
+// ignoreUnknownContext argument must be set to false.
+func NewPriorityContextInferrer(contextsPriority []string, ignoreUnknownContext bool) ContextInferrer {
+	contextPriority := make(map[string]int, len(contextsPriority))
+	for i, ctx := range contextsPriority {
+		contextPriority[ctx] = i
+	}
+	return &priorityContextInferrer{
+		contextPriority:      contextPriority,
+		ignoreUnknownContext: ignoreUnknownContext,
```

Comment:
Erroring on unknown contexts definitely feels like an acceptable outcome. This is a startup situation, so being able to identify a statement like `set(spna.attributes["test"], "pass")` as a typo in `spna` feels really useful.

Reply:
The current implementation is using this flag to [skip](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35721/files#diff-ea1425b770cca2f16c5ecf77b75ed38b7dfeceb03fbc8d5e3cd1bd413ad506d8R45) non-prioritised contexts, it doesn't return an error when it founds those contexts, it either ignores it, or use it with the lowest possible priority. The responsibility of validating either the unknown context is valid or not is delegated to the [parser collection](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/35050/files#diff-7efe3affb096a0fa16edab4c8e48d665bd120c938103b778ad778b15c5f04ce6R190), which is the one that holds all supported contexts. 

The `contextPriority` here acts more like a list of well-defined priorities, and not like a list of valid contexts. One advantage on this approach is that it only needs to know the priorities, and not all (and only) the supported contexts.

Reply:
In that case I think we can probably start with implementing either the skip or "lowest" behavior and adding configurability later if we need it. Whether `spna` is ignored or set as the lowest context, I would want the parser collection (or something) to be able to tell the user that `spna` is an unsupported context.